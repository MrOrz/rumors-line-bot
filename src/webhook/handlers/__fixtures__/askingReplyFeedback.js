export const twoFeedbacks = {
  data: {
    CreateOrUpdateArticleReplyFeedback: {
      feedbackCount: 2,
      reply: { type: 'NOT_ARTICLE' },
    },
  },
};

export const oneFeedback = {
  data: {
    CreateOrUpdateArticleReplyFeedback: {
      feedbackCount: 1,
      reply: { type: 'NOT_ARTICLE' },
    },
  },
};

export const error = {
  errors: [{ message: 'Some error' }],
};
