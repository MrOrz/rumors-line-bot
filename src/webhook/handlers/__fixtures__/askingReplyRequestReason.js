export const createReplyRequestSuccess = {
  data: {
    CreateOrUpdateReplyRequest: {
      replyRequestCount: 1,
    },
  },
};
