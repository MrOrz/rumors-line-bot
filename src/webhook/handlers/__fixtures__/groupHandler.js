const joinGroup = {
  replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
  type: 'join',
  mode: 'active',
  timestamp: 1462629479859,
  source: {
    type: 'group',
    groupId: 'C4a1',
  },
};

const leaveGroup = {
  replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
  type: 'leave',
  mode: 'active',
  timestamp: 1462629479859,
  source: {
    type: 'group',
    groupId: 'C4a1',
  },
};

const textMessage = {
  replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
  type: 'message',
  mode: 'active',
  timestamp: 1462629479859,
  message: {
    type: 'text',
  },
  source: {
    type: 'group',
    groupId: 'C4a1',
  },
};

const expiredTextMessage = {
  replyToken: 'nHuyWiB7yP5Zw52FIkcQobQuGDXCTA',
  type: 'message',
  mode: 'active',
  timestamp: 612921600000,
  message: {
    type: 'text',
  },
  source: {
    type: 'group',
    groupId: 'C4a1',
  },
};

export default { joinGroup, leaveGroup, textMessage, expiredTextMessage };
