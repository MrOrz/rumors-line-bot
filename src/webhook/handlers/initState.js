import stringSimilarity from 'string-similarity';
import { t } from 'ttag';
import gql from 'src/lib/gql';
import {
  createPostbackAction,
  ellipsis,
  createAskArticleSubmissionConsent,
} from './utils';
import ga from 'src/lib/ga';

const SIMILARITY_THRESHOLD = 0.95;

export default async function initState(params) {
  let { data, state, event, userId, replies, isSkipUser } = params;

  // Track text message type send by user
  const visitor = ga(userId, state, event.input);
  visitor.event({ ec: 'UserInput', ea: 'MessageType', el: 'text' });

  // Store user input into context
  data.searchedText = event.input;
  // Store input message type to context for non-init states use
  data.messageType = event.message.type;

  // Search for articles
  const {
    data: { ListArticles },
  } = await gql`
    query($text: String!) {
      ListArticles(
        filter: { moreLikeThis: { like: $text } }
        orderBy: [{ _score: DESC }]
        first: 4
      ) {
        edges {
          node {
            text
            id
          }
        }
      }
    }
  `({
    text: event.input,
  });

  const articleSummary = ellipsis(event.input, 12);

  if (ListArticles.edges.length) {
    // Track if find similar Articles in DB.
    visitor.event({ ec: 'UserInput', ea: 'ArticleSearch', el: 'ArticleFound' });

    // Track which Article is searched. And set tracking event as non-interactionHit.
    ListArticles.edges.forEach(edge => {
      visitor.event({
        ec: 'Article',
        ea: 'Search',
        el: edge.node.id,
        ni: true,
      });
    });

    const edgesSortedWithSimilarity = ListArticles.edges
      .map(edge => {
        edge.similarity = stringSimilarity.compareTwoStrings(
          // Remove spaces so that we count word's similarities only
          //
          edge.node.text.replace(/\s/g, ''),
          event.input.replace(/\s/g, '')
        );
        return edge;
      })
      .sort((edge1, edge2) => edge2.similarity - edge1.similarity)
      .slice(0, 9) /* flex carousel has at most 10 bubbles */;

    // Store article ids
    data.foundArticleIds = edgesSortedWithSimilarity.map(
      ({ node: { id } }) => id
    );

    const hasIdenticalDocs =
      edgesSortedWithSimilarity[0].similarity >= SIMILARITY_THRESHOLD;

    if (edgesSortedWithSimilarity.length === 1 && hasIdenticalDocs) {
      // choose for user
      event.input = 1;

      visitor.send();
      return {
        data,
        state: 'CHOOSING_ARTICLE',
        event,
        userId,
        replies,
        isSkipUser: true,
      };
    }

    const postMessage = edgesSortedWithSimilarity.map(
      ({ node: { text }, similarity }, idx) => {
        const similarityPercentage = Math.round(similarity * 100);
        const similarityEmoji = ['üòê', 'üôÇ', 'üòÄ', 'üòÉ', 'üòÑ'][
          Math.floor(similarity * 4.999)
        ];
        return {
          type: 'bubble',
          direction: 'ltr',
          header: {
            type: 'box',
            layout: 'horizontal',
            spacing: 'md',
            paddingBottom: 'none',
            contents: [
              {
                type: 'text',
                text: similarityEmoji,
                flex: 0,
              },
              {
                type: 'text',
                text: t`Looks ${similarityPercentage}% similar`,
                gravity: 'center',
                size: 'sm',
                weight: 'bold',
                wrap: true,
                color: '#AAAAAA',
              },
            ],
          },
          body: {
            type: 'box',
            layout: 'horizontal',
            spacing: 'none',
            margin: 'none',
            contents: [
              {
                type: 'text',
                text: ellipsis(text, 300, '...'), // 50KB for entire Flex carousel,
                maxLines: 6,
                flex: 0,
                gravity: 'top',
                weight: 'regular',
                wrap: true,
              },
            ],
          },
          footer: {
            type: 'box',
            layout: 'horizontal',
            contents: [
              {
                type: 'button',
                action: createPostbackAction(
                  t`Choose this one`,
                  idx + 1,
                  data.sessionId
                ),
                style: 'primary',
              },
            ],
          },
        };
      }
    );
    postMessage.push({
      type: 'bubble',
      header: {
        type: 'box',
        layout: 'horizontal',
        paddingBottom: 'none',
        contents: [
          {
            type: 'text',
            text: 'üò∂',
            margin: 'none',
            size: 'sm',
            weight: 'bold',
            color: '#AAAAAA',
          },
        ],
      },
      body: {
        type: 'box',
        layout: 'horizontal',
        spacing: 'none',
        margin: 'none',
        contents: [
          {
            type: 'text',
            text: t`None of these messages matches mine :(`,
            maxLines: 5,
            flex: 0,
            gravity: 'top',
            weight: 'regular',
            wrap: true,
          },
        ],
      },
      footer: {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'button',
            action: createPostbackAction(t`Tell us more`, 0, data.sessionId),
            style: 'primary',
          },
        ],
      },
    });

    const templateMessage = {
      type: 'flex',
      altText: t`Please choose the most similar message from the list.`,
      contents: {
        type: 'carousel',
        contents: postMessage,
      },
    };

    const prefixTextArticleFound = [
      {
        type: 'text',
        text: `üîç ${t`There are some messages that looks similar to "${articleSummary}" you have sent to me.`}`,
      },
    ];
    const textArticleFound = [
      {
        type: 'text',
        text:
          t`Internet rumors are often mutated and shared.
            Please choose the version that looks the most similar` + 'üëá',
      },
      templateMessage,
    ];
    if (data.messageType === 'image') {
      replies = textArticleFound;
    } else {
      replies = prefixTextArticleFound.concat(textArticleFound);
    }

    state = 'CHOOSING_ARTICLE';
  } else {
    // Track if find similar Articles in DB.
    visitor.event({
      ec: 'UserInput',
      ea: 'ArticleSearch',
      el: 'ArticleNotFound',
    });

    replies = [
      {
        type: 'text',
        // use `articleSummary` for text only because ocr may get wrong text from image
        text:
          data.messageType === 'text'
            ? t`We didn't find anything about "${articleSummary}" :(`
            : t`We didn't find anything about that :(`,
      },
      ...createAskArticleSubmissionConsent(userId, data.sessionId),
    ];
    state = 'ASKING_ARTICLE_SUBMISSION_CONSENT';
  }
  visitor.send();
  return { data, state, event, userId, replies, isSkipUser };
}
