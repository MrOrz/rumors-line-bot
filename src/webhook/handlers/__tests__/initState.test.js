jest.mock('src/lib/gql');
jest.mock('src/lib/ga');

import MockDate from 'mockdate';
import initState from '../initState';
import * as apiResult from '../__fixtures__/initState';
import gql from 'src/lib/gql';
import ga from 'src/lib/ga';

import { REASON_PREFIX } from 'src/lib/sharedUtils';

beforeEach(() => {
  ga.clearAllMocks();
  gql.__reset();
});

it('article found', async () => {
  gql.__push(apiResult.longArticle);

  const input = {
    data: {
      sessionId: 1497994017447,
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input:
        '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text:
          '計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入',
      },
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "text",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleFound",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "AVvY-yizyCdS-nWhuYWx",
          "ni": true,
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('long article replies still below flex message limit', async () => {
  gql.__push(apiResult.twelveLongArticles);

  const input = {
    data: {
      sessionId: 1502477506267,
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input:
        '這樣的大事國內媒體竟然不敢報導！\n我國駐日代表將原「中華民國」申請更名為「台灣」結果被日本裁罰，須繳納7000萬日圓（合約台幣2100萬元）高額稅賦(轉載中時電子報）\n\n我駐日代表謝長廷將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭自認友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元）高額稅賦...民進黨沒想到如此更名竟然是這樣的下場：被他最信任也最友好的日本政府給坑了。\n果然錯誤的政策比貪污可怕，2100萬就這樣打水漂了，還要資助九州水患，核四停建違約賠償金.......夠全國軍公教退休2次.........\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118',
      timestamp: 1502477505309,
      message: {
        type: 'text',
        id: '6530038889933',
        text:
          '這樣的大事國內媒體竟然不敢報導！\n我國駐日代表將原「中華民國」申請更名為「台灣」結果被日本裁罰，須繳納7000萬日圓（合約台幣2100萬元）高額稅賦(轉載中時電子報）\n\n我駐日代表謝長廷將原「中華民國」申請更名為「台灣」，自認得意之時，結果遭自認友好日本國給出賣了，必須繳納7000萬日圓（合約台幣2100萬元）高額稅賦...民進黨沒想到如此更名竟然是這樣的下場：被他最信任也最友好的日本政府給坑了。\n果然錯誤的政策比貪污可怕，2100萬就這樣打水漂了，還要資助九州水患，核四停建違約賠償金.......夠全國軍公教退休2次.........\n\nhttp://www.chinatimes.com/newspapers/20170617000318-260118',
      },
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    isSkipUser: false,
  };

  const result = await initState(input);
  expect(gql.__finished()).toBe(true);
  expect(result.replies.length).toBeLessThanOrEqual(5); // Reply message API limit
  const carousel = result.replies.find(({ type }) => type === 'flex').contents;
  expect(carousel.type).toBe('carousel');
  expect(carousel.contents.length).toBeLessThanOrEqual(10); // Flex message carousel 10 bubble limit
  expect(JSON.stringify(carousel).length).toBeLessThan(50 * 1000); // Flex message carousel 50K limit
});

it('articles found with high similarity', async () => {
  gql.__push(apiResult.twoShortArticles);

  const input = {
    data: {
      sessionId: 1497994017447,
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input: 'YouTube · 寻找健康人生',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text: 'YouTube · 寻找健康人生',
      },
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "text",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleFound",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "AVvY-yizyCdS-nWhuYWx",
          "ni": true,
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "AVvY-yizyCdS-nWhuYWy",
          "ni": true,
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('only one article found with high similarity', async () => {
  gql.__push(apiResult.shortArticle);

  const input = {
    data: {
      sessionId: 1497994017447,
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input: 'YouTube · 寻找健康人生',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text: 'YouTube · 寻找健康人生',
      },
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "text",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleFound",
        },
      ],
      Array [
        Object {
          "ea": "Search",
          "ec": "Article",
          "el": "AVvY-yizyCdS-nWhuYWx",
          "ni": true,
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('should handle text not found', async () => {
  gql.__push(apiResult.notFound);

  const input = {
    data: {
      sessionId: 1497994017447,
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input:
        'YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...',
      timestamp: 1497994016356,
      message: {
        type: 'text',
        id: '6270464463537',
        text:
          'YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...',
      },
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  MockDate.set('2020-01-01');
  expect(await initState(input)).toMatchSnapshot();
  MockDate.reset();
  expect(gql.__finished()).toBe(true);

  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "text",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleNotFound",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('should handle image not found', async () => {
  gql.__push(apiResult.notFound);

  const input = {
    data: {
      sessionId: 1497994017447,
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input: 'OCR text OCR text OCR text',
      timestamp: 1497994016356,
      message: {
        type: 'image',
      },
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await initState(input)).toMatchSnapshot();
  expect(gql.__finished()).toBe(true);

  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "MessageType",
          "ec": "UserInput",
          "el": "image",
        },
      ],
      Array [
        Object {
          "ea": "ArticleSearch",
          "ec": "UserInput",
          "el": "ArticleNotFound",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('handles reason LIFF: incorrect context', async () => {
  const input = {
    data: {
      sessionId: 1497994017447,
      // No selectedArticleId
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input: REASON_PREFIX + 'My reason',
      timestamp: 1497994016356,
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  await expect(initState(input)).rejects.toMatchInlineSnapshot(
    `[Error: Please press the latest button to submit message to database.]`
  );
  expect(ga.sendMock).toHaveBeenCalledTimes(0);
});

it('handles reason LIFF: reply request update failed', async () => {
  const input = {
    data: {
      sessionId: 1497994017447,
      selectedArticleId: 'article-id',
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input: REASON_PREFIX + 'My reason',
      timestamp: 1497994016356,
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  gql.__push(apiResult.apiError);

  await expect(initState(input)).rejects.toMatchInlineSnapshot(
    `[Error: Something went wrong when recording your reason, please try again later.]`
  );
  expect(gql.__finished()).toBe(true);
  expect(ga.sendMock).toHaveBeenCalledTimes(0);
});

it('handles reason LIFF: reply request update success, only 1 reply request', async () => {
  const input = {
    data: {
      sessionId: 1497994017447,
      selectedArticleId: 'article-id',
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input: REASON_PREFIX + 'My reason',
      timestamp: 1497994016356,
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  gql.__push(apiResult.updateReplyRequestSuccess);

  MockDate.set('2020-01-01');
  expect(await initState(input)).toMatchSnapshot();
  MockDate.reset();

  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "ProvidingReason",
          "ec": "Article",
          "el": "article-id",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});

it('handles reason LIFF: reply request update success, multiple reply requests', async () => {
  const input = {
    data: {
      sessionId: 1497994017447,
      selectedArticleId: 'article-id',
    },
    state: '__INIT__',
    event: {
      type: 'message',
      input: REASON_PREFIX + 'My reason',
      timestamp: 1497994016356,
    },
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  gql.__push(apiResult.updateReplyRequestSuccess2);

  MockDate.set('2020-01-01');
  expect(await initState(input)).toMatchSnapshot();
  MockDate.reset();

  expect(gql.__finished()).toBe(true);
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "ProvidingReason",
          "ec": "Article",
          "el": "article-id",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});
