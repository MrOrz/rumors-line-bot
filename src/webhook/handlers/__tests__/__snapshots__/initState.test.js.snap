// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article found 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
    ],
    "messageType": "text",
    "searchedText": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
    "sessionId": 1497994017447,
  },
  "event": Object {
    "input": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
    "message": Object {
      "id": "6270464463537",
      "text": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "replies": Array [
    Object {
      "text": "🔍 There are some messages that looks similar to \\"計程車上有裝悠遊卡感⋯⋯\\" you have sent to me.",
      "type": "text",
    },
    Object {
      "text": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
      "type": "text",
    },
    Object {
      "altText": "Please choose the most similar message from the list.",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 6,
                  "text": "計程車上有裝悠遊卡感應器，老人悠悠卡可以享受優惠部分由政府補助，不影響司機收入，下車時使用老人悠遊卡，跳錶車資105元，優惠32元，只扣73元，哈哈，這是屬於我們的福利，與大家分享，可以善加利用！=7折，朋友使用ok",
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "margin": "none",
              "spacing": "none",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":1,\\"sessionId\\":1497994017447}",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😀",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "Looks 52% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 5,
                  "text": "None of these messages matches mine :(",
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "margin": "none",
              "spacing": "none",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":0,\\"sessionId\\":1497994017447}",
                    "label": "Tell us more",
                    "type": "postback",
                  },
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "margin": "none",
                  "size": "sm",
                  "text": "😶",
                  "type": "text",
                  "weight": "bold",
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`articles found with high similarity 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
      "AVvY-yizyCdS-nWhuYWy",
    ],
    "messageType": "text",
    "searchedText": "YouTube · 寻找健康人生",
    "sessionId": 1497994017447,
  },
  "event": Object {
    "input": "YouTube · 寻找健康人生",
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "replies": Array [
    Object {
      "text": "🔍 There are some messages that looks similar to \\"YouTube · ⋯⋯\\" you have sent to me.",
      "type": "text",
    },
    Object {
      "text": "Internet rumors are often mutated and shared.
Please choose the version that looks the most similar👇",
      "type": "text",
    },
    Object {
      "altText": "Please choose the most similar message from the list.",
      "contents": Object {
        "contents": Array [
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 6,
                  "text": "YouTube  ·  寻找健康人生",
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "margin": "none",
              "spacing": "none",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":1,\\"sessionId\\":1497994017447}",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 6,
                  "text": "YouTube
·
寻找健康人生",
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "margin": "none",
              "spacing": "none",
              "type": "box",
            },
            "direction": "ltr",
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":2,\\"sessionId\\":1497994017447}",
                    "label": "Choose this one",
                    "type": "postback",
                  },
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "text": "😄",
                  "type": "text",
                },
                Object {
                  "color": "#AAAAAA",
                  "gravity": "center",
                  "size": "sm",
                  "text": "Looks 100% similar",
                  "type": "text",
                  "weight": "bold",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "spacing": "md",
              "type": "box",
            },
            "type": "bubble",
          },
          Object {
            "body": Object {
              "contents": Array [
                Object {
                  "flex": 0,
                  "gravity": "top",
                  "maxLines": 5,
                  "text": "None of these messages matches mine :(",
                  "type": "text",
                  "weight": "regular",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "margin": "none",
              "spacing": "none",
              "type": "box",
            },
            "footer": Object {
              "contents": Array [
                Object {
                  "action": Object {
                    "data": "{\\"input\\":0,\\"sessionId\\":1497994017447}",
                    "label": "Tell us more",
                    "type": "postback",
                  },
                  "style": "primary",
                  "type": "button",
                },
              ],
              "layout": "horizontal",
              "type": "box",
            },
            "header": Object {
              "contents": Array [
                Object {
                  "color": "#AAAAAA",
                  "margin": "none",
                  "size": "sm",
                  "text": "😶",
                  "type": "text",
                  "weight": "bold",
                },
              ],
              "layout": "horizontal",
              "paddingBottom": "none",
              "type": "box",
            },
            "type": "bubble",
          },
        ],
        "type": "carousel",
      },
      "type": "flex",
    },
  ],
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`only one article found with high similarity 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVvY-yizyCdS-nWhuYWx",
    ],
    "messageType": "text",
    "searchedText": "YouTube · 寻找健康人生",
    "sessionId": 1497994017447,
  },
  "event": Object {
    "input": 1,
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": true,
  "replies": undefined,
  "state": "CHOOSING_ARTICLE",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should handle article not found 1`] = `
Object {
  "data": Object {
    "messageType": "text",
    "searchedText": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
    "sessionId": 1497994017447,
  },
  "event": Object {
    "input": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
    "message": Object {
      "id": "6270464463537",
      "text": "YouTube · 寻找健康人生 驚！大批香蕉受到愛滋血污染！這種香蕉千萬不要吃！吃到可能會被 ...",
      "type": "text",
    },
    "timestamp": 1497994016356,
    "type": "message",
  },
  "isSkipUser": false,
  "replies": Array [
    Object {
      "text": "We didn't find anything about \\"YouTube · ⋯⋯\\" :(",
      "type": "text",
    },
    Object {
      "altText": "🥇 Be the first to report the message
📱 Please proceed on your mobile phone.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "contents": Array [
                Object {
                  "text": "Currently we don't have this message in our database. If you think it is probably a rumor, ",
                  "type": "span",
                },
                Object {
                  "color": "#ffb600",
                  "text": "press “🆕 Submit to database” to make this message public on Cofacts database ",
                  "type": "span",
                  "weight": "bold",
                },
                Object {
                  "text": "for nice volunteers to fact-check. Although you won't receive answers rightaway, you can help the people who receive the same message in the future.",
                  "type": "span",
                },
              ],
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "paddingAll": "lg",
          "spacing": "md",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "label": "🆕 Submit to database",
                "type": "uri",
                "uri": "line://app/1631030389-xb5mnDdN?p=source&token=eyJhbGciOiJIUzI1NiJ9.eyJzZXNzaW9uSWQiOjE0OTc5OTQwMTc0NDcsImV4cCI6MTU3NzkyMzIwMH0.swCHAII1dNMjmwCyhIIBZII_Csw8ME_xe5U6TgjNRJQ",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "header": Object {
          "contents": Array [
            Object {
              "flex": 0,
              "gravity": "center",
              "text": "🥇",
              "type": "text",
            },
            Object {
              "color": "#ffb600",
              "text": "Be the first to submit the message",
              "type": "text",
              "weight": "bold",
              "wrap": true,
            },
          ],
          "layout": "horizontal",
          "paddingAll": "lg",
          "spacing": "sm",
          "type": "box",
        },
        "styles": Object {
          "body": Object {
            "separator": true,
          },
        },
        "type": "bubble",
      },
      "type": "flex",
    },
  ],
  "state": "ASKING_ARTICLE_SUBMISSION_CONSENT",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;
